apps/aecore/src/aec_chain_state.erl

```diff
@@ -151,19 +163,28 @@ node_is_genesis(Node, #{genesis_block_hash := undefined}) ->
 node_is_genesis(Node, State) ->
     hash(Node) =:= get_genesis_hash(State).
 
-wrap_block(Block) ->
+wrap_block(Block, Type) ->
     Header = aec_blocks:to_header(Block),
     {ok, Hash} = aec_headers:hash_header(Header),
     #node{ header = Header
          , hash = Hash
+         , height = aec_headers:height(Header)
+         , type = Type
          }.
 
 wrap_header(Header) ->
     {ok, Hash} = aec_headers:hash_header(Header),
     #node{ header = Header
          , hash = Hash
+         , height = aec_headers:height(Header)
+         , type = aec_blocks:type()
          }.
 
+set_height(#node{type = micro} = Node, #{top_key_block_height := Height}) ->
```

Comment:
This is not safe. You don't know that this micro block is at the top. It might even be on a different fork.

If the height is supposed to be inherited from the keyblock, you should check that there is a chain of microblocks to a keyblock and then get the height from there. I don't see why the height is not given in the micro blocks, and then is checked to match in a micro chain back to the key block.